---
title: "Metabarcoding Singapore <br> Process ASV from Dada2"
author: "Daniel Vaulot, Adriana Lopes dos Santos"
date: '`r format(Sys.time(), "%d %m %Y")`'
header-includes:
   - \usepackage{color, fancyvrb}
output:
  rmdformats::readthedown:
    highlight: kate
    toc_depth: 3
    number_sections : yes    
  pdf_document: 
    toc: yes
    toc_depth: 3
    number_sections : yes
---


# Aim

* Assign and analyze eukaryotes for Singapore metabarcoding data (ASV assigned with dada2 as implemented on Mothur). 



# Initialize

This file defines all the necessary libraries and variables

```{r init, eval=TRUE, message=FALSE, warning=FALSE}
  source('Metabarcoding Singapore_init.R', echo=FALSE)
```


# Read the data
## File names

```{r}
  full_path_data <- function(file_name) { str_c("../qiime/2018-09-06_dada2/", file_name)}

  taxo_file <- full_path_data("taxonomy.tsv")
  otu_file <- full_path_data("feature-table_unrarefied.tsv")
  otu_excel_file <- "../qiime/Singapore ASV_table.xlsx"
  sequence_file <- full_path_data("ref_sequences_unrarefied.fasta")
  
  metadata_xlsx <- "../metadata/Singapore_metadata.xlsx"
  metadata_csv <- "../metadata/monsoonpaper_env_data.csv"
  
  sequence_file_euk <- full_path_data("ASV_unrarefied_euk.fasta")
  sequence_file_mamiello <- full_path_data("ASV_Mamiello.fasta")
  
  dada2_taxo_file_euk <- full_path_data("ASV_unrarefied_euk.dada2.taxo")
  dada2_boot_file_euk <- full_path_data("ASV_unrarefied_euk.dada2.boot")
  otu_table_final_file <- full_path_data("ASV_final.tsv")
  
  blast_file <- full_path_data("ASV_unrarefied_euk.blast.tsv")  
```


## Read the files
* The dada2 treatment has already removed the forward and reverse primers, so no need to remove them
* Work with the unrarefied data


```{r}
# Read the sample and metadata tables
  sample_table <- read_excel(metadata_xlsx, sheet="samples", range="A1:D89") 

  metadata_table  <- read_csv(metadata_csv, na=c("ND", "")) %>% 
    dplyr::rename(sample_code=Sample, day_number=Day_number, date=Date, 
                  location=Location, monsoon=`Monsoon period`) %>% 
    select(-Strait) %>% 
    mutate(date = lubridate::parse_date_time(date,"dmy"),
           monsoon = forcats::fct_relevel(monsoon, "NE", "IM-1", "SW", "IM-2"))
  
  station_table  <- read_excel(metadata_xlsx, sheet="stations", na=c("ND", ""))

  sample_table <- sample_table %>% 
    left_join(metadata_table) %>%
    left_join(station_table) %>% 
    mutate(sample_label = str_c(strait_label,location_label,
                                monsoon,sprintf("%03d",day_number), 
                                sep="_"))

# Read the taxonomy table
  taxo_table <- read_tsv(taxo_file)

# Clean up the taxonomy
  taxo_table <- taxo_table %>% 
    mutate(taxo_clean = str_replace_all(Taxon, "D_[0-9]+__","")) %>% 
    separate(col=taxo_clean, into=str_c("taxo", c(0:6)), sep=";") %>% 
    rename(otu_name = `Feature ID`)
  
# Remove duplicate entries for bacteria
  pattern_taxa_removed <- "marine metagenome|uncultured"
 
  taxo_table <- taxo_table %>% 
    mutate(taxo2 = case_when (
                      str_detect(taxo2,pattern_taxa_removed) ~ NA_character_,
                      TRUE ~ taxo2),
           taxo3 = case_when (
                      str_detect(taxo3,pattern_taxa_removed) ~ NA_character_,
                      TRUE ~ taxo3),
           taxo4 = case_when (
                      str_detect(taxo4,pattern_taxa_removed) ~ NA_character_,
                      TRUE ~ taxo4),
           taxo5 = case_when (
                      str_detect(taxo5,pattern_taxa_removed) ~ NA_character_,
                      TRUE ~ taxo5),
           taxo6 = case_when (
                      str_detect(taxo6,pattern_taxa_removed) ~ NA_character_,
                      TRUE ~ taxo6))
  
# If taxo_i is empty, fill with taxo_i-1
  taxo_levels <- str_c("taxo", c(0:6))
  taxo_levels_number <- length(taxo_levels)

  for (i in 1:nrow(taxo_table)) {
    for (j in 1:(taxo_levels_number) ){
      if (is.na(taxo_table[i,taxo_levels[j]])) {
              taxo_table[i,taxo_levels[j]] <-   taxo_table[i,taxo_levels[j-1]]
          }
    }    
  }

# Read the otu table
  otu_table <- read_tsv(otu_file, skip=1) %>%  # Jump the first line
    rename(otu_name = `#OTU ID`) %>% 
    mutate(otu_id = str_c("otu_", sprintf("%04d",row_number())))

# Read the sequences
  otu_sequences <- readAAStringSet(sequence_file)
  otu_sequences.df <- data.frame (otu_name=names(otu_sequences),sequence=as.character(otu_sequences))

# Remove the primers - Not necessary because the primers have been removed  
  # fwd_length = 20
  # rev_length = 15
  # otu_sequences.df <- otu_sequences.df %>% 
  #   separate (col=names, into=c("otu_id_qiime", "otu_rep_seq"), sep=" ") %>%
  #   mutate (sequence = str_sub(sequence, start=fwd_length+1, end = - rev_length - 1))

  otu_table <- taxo_table %>%  
    left_join(otu_table) %>% 
    left_join(otu_sequences.df) %>% 
    arrange(otu_id)
  
# Write a fasta file for blast with all taxonomy roups
 # otu_sequences <- otu_table %>% transmute(sequence=sequence, seq_name=otu_id)
 # fasta_write(otu_sequences, file_name="../qiime/otu_rep_98_all.fasta", compress = FALSE, taxo_include = FALSE)  
# write_tsv(otu_table, full_path_data("otu_table.tsv"), na="")
  
```

## Only keep the eukaryotes in the OTU file

```{r}
 otu_table_euk <- otu_table %>% filter(str_detect(Taxon, "Eukaryota"))

 # Write the fasta file file
 otu_sequences_euk <- otu_table_euk %>% transmute(sequence=sequence, seq_name=otu_id)
 fasta_write(otu_sequences_euk, file_name=sequence_file_euk, 
             compress = FALSE, taxo_include = FALSE)
 
```

# Assignment of eukaryotic ASVs based on PR2 database

## Use dada2 to reassign to PR2

```{r, eval=FALSE}
 dada2_assign(seq_file_name=sequence_file_euk,
  ref_file_name = "C:/daniel.vaulot@gmail.com/Databases/_PR2/versions/4.11.0/pr2_version_4.11.0_dada2.fasta.gz")
 
```

## Read the PR2 assignement and merge with initial otu table

```{r}
 otu_euk_pr2 <- read_tsv(dada2_taxo_file_euk) %>% 
   rename(otu_id= seq_name)
 # otu_euk_pr2_boot <- read_tsv(dada2_boot_file_euk) %>%
 #   rename_all(funs(str_c(.,"_boot"))) %>% 
 #   rename(seq_name = seq_name_boot)
 # otu_euk_pr2 <- left_join(otu_euk_pr2, otu_euk_pr2_boot) %>% 
 #   rename(otu_id= seq_name)
 otu_table_final <- left_join(otu_table, otu_euk_pr2) %>% 
   select(otu_id, otu_name,taxo0:taxo6, Taxon, kingdom:species_boot,matches("EC|PR|RM|SBW|STJ"),sequence)
          
 # write_tsv(otu_table_final, otu_table_final_file, na = "")
```

## Export fasta file with taxonomy for Mamiellophyceae

```{r}
  otu_mamiello <- otu_table_final %>% 
    filter(class == "Mamiellophyceae") %>% 
    select(sequence=sequence, seq_name=otu_id, supergroup:species)
 fasta_write(otu_mamiello, file_name=sequence_file_mamiello, 
             compress = FALSE, taxo_include = TRUE)

```

# Read reassignments for some prokaryotes

## Synechococcus

```{r}
 otu_table_syn <- read_excel(otu_excel_file, sheet="syn_reassigned")
 otu_table_final <- otu_table_final %>% filter(!str_detect(taxo5, "Synechococcus")) %>% 
   bind_rows(otu_table_syn)
```

# Process BLAST file

BLAST is performed on Roscoff ABIMS server

```{r, eval=FALSE}
blast_18S_reformat(blast_file)
```

