---
title: "Metabarcoding Singapore <br> OTUs at 98%"
author: "Daniel Vaulot, Adriana Lopes dos Santos"
date: '`r format(Sys.time(), "%d %m %Y")`'
header-includes:
   - \usepackage{color, fancyvrb}
output:
  rmdformats::readthedown:
    highlight: kate
    toc_depth: 3
    number_sections : yes    
  pdf_document: 
    toc: yes
    toc_depth: 3
    number_sections : yes
---


Initialize. This file defines all the necessary libraries and variables

```{r init, eval=TRUE, message=FALSE, warning=FALSE}
  source('Metabarcoding Singapore_init.R', echo=FALSE)
```

# Aim

* Assign and analyze eukaryotes for Singapore metabarcoding data (98%).  Do some analyzes with the prokaryotes too...

# Analysis

## Assignment of 98% OTUs based on PR2 database

### Read the data

```{r}
# Read the sample table
  sample_table <- read_excel("../qiime/Singapore otu_table_w_tax_0.98.xlsx", sheet="samples") %>% 
                  mutate(sample_label = str_c(strait,site_M,sprintf("%03d",day), sep="_"))

# Read the otu table
  otu_table <- read_excel("../qiime/Singapore otu_table_w_tax_0.98.xlsx", sheet="otu_table")

# Clean up the taxonomy
  otu_table <- otu_table %>% 
    mutate(taxo_clean = str_replace_all(taxonomy, "D_[0-9]+__","")) %>% 
    separate(col=taxo_clean, into=str_c("taxo", c(1:7)), sep="; ")

# Read the sequences
  otu_sequences <- readAAStringSet("../qiime/otu_rep_98.fna")
  otu_sequences.df <- data.frame (names=names(otu_sequences),sequence=as.character(otu_sequences))

# Split the otu sequences name into the otu id given by qiime and the name of the representative sequence  
# Remove the primers 
  fwd_length = 20
  rev_length = 15
  otu_sequences.df <- otu_sequences.df %>% 
    separate (col=names, into=c("otu_id_qiime", "otu_rep_seq"), sep=" ") %>%
    mutate (sequence = str_sub(sequence, start=fwd_length+1, end = - rev_length - 1))
  otu_table <- left_join(otu_table,otu_sequences.df)
  
 # Write a fasta file for blast with all taxonomy roups
 otu_sequences <- otu_table %>% transmute(sequence=sequence, seq_name=otu_id)
 fasta_write(otu_sequences, file_name="../qiime/otu_rep_98_all.fasta", compress = FALSE, taxo_include = FALSE)  
  
  
```

### Only keep the eukaryotes in the OTU file

```{r}
 otu_table_euk <- otu_table %>% filter(str_detect(taxonomy, "Eukaryota|Unassigned"))

 # Write the fasta file file
 otu_sequences_euk <- otu_table_euk %>% transmute(sequence=sequence, seq_name=otu_id)
 fasta_write(otu_sequences_euk, file_name="../qiime/otu_rep_98_euk.fasta", compress = FALSE, taxo_include = FALSE)
 
```
### Use dada2 to reassign to PR2

```{r, eval=FALSE}
 dada2_assign(seq_file_name="../qiime/otu_rep_98_euk.fasta")
 
```

### Read the PR2 assignement and merge with initial otu table

```{r}
 otu_euk_pr2 <- read_tsv("../qiime/otu_rep_98_euk.dada2.taxo")
 otu_euk_pr2_boot <- read_tsv("../qiime/otu_rep_98_euk.dada2.boot") %>%
   rename_all(funs(str_c(.,"_boot"))) %>% 
   rename(seq_name = seq_name_boot)
 otu_euk_pr2 <- left_join(otu_euk_pr2, otu_euk_pr2_boot) %>% 
   rename(otu_id= seq_name)
 otu_table_final <- left_join(otu_table, otu_euk_pr2) %>% 
   select(otu_id, otu_id_qiime,taxo1:taxo7, kingdom:species_boot,nseq, STJ11.S23:PR06.S10, otu_rep_seq, sequence)
          
 write_tsv(otu_table_final, "../qiime/otu_table_98_final.tsv")
```


## Phyloseq analysis

### Create phyloseq files for euk after filtering the data

Filter the euk data to remove the low bootstraps values (threshold : bootstrap > 90% at the supergroup level) and create a phyloseq file

Note the bootstrap threshold had to be higher for 98% compared to 97% (90% vs 65%)
```{r}
  otu_table_euk_final <- otu_table_final %>% filter(supergroup_boot > 90)
  str_c("Number of euk otus : ", nrow(otu_table_euk_final))

  otu_mat<- otu_table_euk_final %>%  select(otu=otu_id,STJ11.S23:PR06.S10 )
  tax_mat<- otu_table_euk_final %>%  select(otu=otu_id,kingdom:species)
  samples_df <- sample_table %>%  rename (sample=sample_id)
  
  row.names(otu_mat) <- otu_mat$otu
  otu_mat <- otu_mat %>% select (-otu) 
  row.names(tax_mat) <- tax_mat$otu
  tax_mat <- tax_mat %>% select (-otu) 
  row.names(samples_df) <- samples_df$sample
  samples_df <- samples_df %>% select (-sample)
  otu_mat <- as.matrix(otu_mat)
  tax_mat <- as.matrix(tax_mat)
  
  OTU = otu_table(otu_mat, taxa_are_rows = TRUE)
  TAX = tax_table(tax_mat)
  samples = sample_data(samples_df)
  
  singa_euk <- phyloseq(OTU, TAX, samples)
  singa_euk
```
Break up into photosynthetic and non-photosynthetic (Metazoa are removed)

```{r}
  singa_photo <- subset_taxa(singa_euk, (division %in% c("Chlorophyta", "Dinoflagellata", "Cryptophyta", "Rhodophyta",
                                                            "Haptophyta", "Ochrophyta", "Cercozoa")) &
                                         !(class %in% c("Syndiniales", "Sarcomonadea")))
  singa_photo
  
  
  singa_hetero <- subset_taxa(singa_euk,!(division %in% c("Chlorophyta", "Dinoflagellata", "Cryptophyta","Rhodophyta", 
                                                            "Haptophyta", "Ochrophyta", "Cercozoa", "Metazoa")) |
                                          (class %in% c("Syndiniales", "Sarcomonadea")))
 
  singa_hetero
```


Normalize number of reads in each sample using median sequencing depth.

```{r}
  total_photo = median(sample_sums(singa_photo))
  sprintf("The number of reads used for normalization of autotrophs is  %.0f", total_photo)
  standf = function(x, t=total_photo) (t * (x / sum(x)))
  singa_photo = transform_sample_counts(singa_photo, standf)
  
  total_hetero = median(sample_sums(singa_hetero))
  sprintf("The number of reads used for normalization of heterotrophs is  %.0f", total_hetero)
  standf = function(x, t=total_hetero) (t * (x / sum(x)))
  singa_hetero = transform_sample_counts(singa_hetero, standf)

```
Remove taxa which are in low abundace (< 0.10 in any given sample) and normalize again...

```{r}
  contrib_min <- 0.1

# Photo abundant
  singa_photo_abundant <- filter_taxa(singa_photo, function(x) sum(x > total_photo*contrib_min) > 0, TRUE)
  
  total = median(sample_sums(singa_photo_abundant))
  sprintf("The number of reads used for normalization of abundant autotrophs is  %.0f", total)
  standf = function(x, t=total) (t * (x / sum(x)))
  singa_photo_abundant = transform_sample_counts(singa_photo_abundant, standf)
 
  singa_photo_abundant
  
# Hetero abundant  
  singa_hetero_abundant <- filter_taxa(singa_hetero, function(x) sum(x > total_hetero*contrib_min) > 0, TRUE)
  
  total = median(sample_sums(singa_hetero_abundant))
  sprintf("The number of reads used for normalization of abundant heterotrophs is  %.0f", total)
  standf = function(x, t=total) (t * (x / sum(x)))
  singa_hetero_abundant = transform_sample_counts(singa_hetero_abundant, standf)

  singa_hetero_abundant
  
  
```


### Create phyloseq files for proks

```{r}
  otu_table_prok <- otu_table %>% filter(taxo1 %in% c("Bacteria", "Archaea"))
  
  otu_mat<- otu_table_prok %>%  select(otu=otu_id,STJ11.S23:PR06.S10 )
  
  tax_mat<- otu_table_prok %>%  select(otu=otu_id,taxo1:taxo7) %>% 
    rename(kingdom=taxo1, supergroup=taxo2, division=taxo3, class=taxo4, order=taxo5, family=taxo6, genus=taxo7) %>% 
    mutate(species=NA)
  
  samples_df <- sample_table %>%  rename (sample=sample_id)
  
  row.names(otu_mat) <- otu_mat$otu
  otu_mat <- otu_mat %>% select (-otu) 
  row.names(tax_mat) <- tax_mat$otu
  tax_mat <- tax_mat %>% select (-otu) 
  row.names(samples_df) <- samples_df$sample
  samples_df <- samples_df %>% select (-sample)
  otu_mat <- as.matrix(otu_mat)
  tax_mat <- as.matrix(tax_mat)
  
  OTU = otu_table(otu_mat, taxa_are_rows = TRUE)
  TAX = tax_table(tax_mat)
  samples = sample_data(samples_df)
  
  singa_prok <- phyloseq(OTU, TAX, samples)
```

Normalize number of reads in each sample using median sequencing depth.

```{r}
  total_prok = median(sample_sums(singa_prok))
  sprintf("The number of reads used for normalization of prokaryotes is  %.0f", total_prok)
  standf = function(x, t=total_prok) (t * (x / sum(x)))
  singa_prok= transform_sample_counts(singa_prok, standf)
  
  singa_prok

```
Remove otus which are in low abundace (< 0.05 in any given sample) and normalize again

```{r}
  contrib_min <- 0.05
  singa_prok_abundant <- filter_taxa(singa_prok, function(x) sum(x > total_prok*contrib_min) > 0, TRUE)
  total = median(sample_sums(singa_prok_abundant))
  sprintf("The number of reads used for normalization of abundant prokaryotes is  %.0f", total)
  standf = function(x, t=total) (t * (x / sum(x)))
  singa_prok_abundant= transform_sample_counts(singa_prok_abundant, standf)
  
  singa_prok_abundant
  
```


### Create a list for the auto and hetero phyloseq files

```{r}
  ps_list <- list(ps=c(singa_photo, singa_hetero, singa_prok), title = c("Eukaryotes - Autotrophs - all OTUs", "Eukaryotes - Heterotrophs - all OTUs", "Prokaryotes - all OTUs"))
  ps_list_abundant <- list(ps=c(singa_photo_abundant, singa_hetero_abundant, singa_prok_abundant), title = c("Eukaryotes - Autotrophs - abundant OTUs (> 10%)", "Eukaryotes - Heterotrophs - abundant OTUs (> 10%)", "Prokaryotes - abundant OTUs (> 5%)"))
```


### Bar plot of divisions per station for auto and hetero

```{r }

for (i in 1:3) {
  p <- plot_bar(ps_list$ps[[i]], x="sample_label", fill = "division") + 
       geom_bar(aes(color=division, fill=division), stat="identity", position="stack") + 
       ggtitle(str_c("Division level - ",ps_list$title[[i]]))  +
       theme(axis.text.x = element_text(angle = 90, vjust=0.5, hjust=1))
  print(p)
}
```

### Bar plot of class per station for auto and hetero

Only consider the abundant taxa

```{r }

for (i in 1:3) {
  p <- plot_bar(ps_list_abundant$ps[[i]], x="sample_label", fill = "class") + 
       geom_bar(aes(color=class, fill=class), stat="identity", position="stack") + 
       ggtitle(str_c("Class level - ",ps_list_abundant$title[[i]]))  +
       theme(axis.text.x = element_text(angle = 90, vjust=0.5, hjust=1))
  print(p)
}
```

### Compare by Straight, Site, Moonsoon (abundant OTUs only)

```{r}
for (factor in c("strait", "site", "moonsoon")) {  
 for (i in 1:3) {
  ps_aggregate <- merge_samples(ps_list_abundant$ps[[i]], factor)
  p <- plot_bar(ps_aggregate, fill = "division") + 
      geom_bar(aes(color=division, fill=division), stat="identity", position="stack") +
      ggtitle(str_c(ps_list_abundant$title[[i]], " - ", factor)) +
      theme(axis.text.x = element_text(angle = 90, vjust=0.5, hjust=1))
  print(p)
 }
}
```


### Main genera for different division of Eukaryotes (Autotrophs)

```{r }
for (one_division in c("Chlorophyta", "Dinoflagellata", "Ochrophyta")){
  ps_subset <- subset_taxa(singa_photo_abundant, division %in% one_division)
  p <- plot_bar(ps_subset, x="genus", fill = "genus", facet_grid = strait~moonsoon) +
    geom_bar(aes(color=genus, fill=genus), stat="identity", position="stack") +
    theme(axis.text.x = element_text(angle = 90, vjust=0.5, hjust=1)) +
    ggtitle(str_c(one_division," - Abundant OTUs"))
  print(p)  
}
```

### Main species of Mamiellophyceae and Diatoms (Eukaryotes - Autrotrophs)

```{r }
for (one_class in c("Mamiellophyceae", "Bacillariophyta")){
  ps_subset <- subset_taxa(singa_photo_abundant, class %in% one_class)
  p <- plot_bar(ps_subset, x="species", fill = "species", facet_grid = strait~moonsoon) +
    geom_bar(aes(color=species, fill=species), stat="identity", position="stack") +
    theme(axis.text.x = element_text(angle = 90, vjust=0.5, hjust=1)) +
    ggtitle(str_c(one_class," - Abundant OTUs"))
  print(p)  
}
```

### Heatmaps (abundant OTUs)

```{r}

for (i in 1:2) {

  p <- plot_heatmap(ps_list_abundant$ps[[i]], method = "NMDS", distance = "bray", 
               taxa.label = "species",  taxa.order="division",
               sample.label = "sample_label", sample.order="sample_label",
               low="beige", high="red", na.value="beige", 
               title=ps_list_abundant$title[[i]])
  print(p)
}

for (i in c(3)) {

  p <- plot_heatmap(ps_list_abundant$ps[[i]], method = "NMDS", distance = "bray", 
               taxa.label = "family",  taxa.order="division",
               sample.label = "sample_label", sample.order="sample_label",
               low="beige", high="red", na.value="beige", 
               title=ps_list_abundant$title[[i]])
  print(p)
}
  
```

### NMDS

All OTUs
```{r}
for (i in 1:3) {
  
  singa.ord <- ordinate(ps_list$ps[[i]], "NMDS", "bray")
  print(singa.ord)

  p <- plot_ordination(ps_list$ps[[i]], singa.ord, type="samples", color="moonsoon", 
                  shape="strait", title=ps_list$title[[i]]) +
       geom_point(size=3) +
       scale_colour_manual(values=c("yellow", "red","blue"))

  print(p)
}
```

Abundant OTUs
```{r}
for (i in 1:3) {
  
  singa.ord <- ordinate(ps_list_abundant$ps[[i]], "NMDS", "bray")
  print(singa.ord)

  p <- plot_ordination(ps_list$ps[[i]], singa.ord, type="samples", color="moonsoon", 
                  shape="strait", title=ps_list_abundant$title[[i]]) +
       geom_point(size=3)  +
       scale_colour_manual(values=c("yellow", "red","blue"))
  print(p)
  
  p <- plot_ordination(ps_list$ps[[i]], singa.ord, type="taxa", color="division", 
                  shape="division", title=ps_list_abundant$title[[i]]) +
       geom_point(size=3)
  print(p)
}
```

### Network analysis

```{r fig.height=6, fig.width=10}

for (i in 1:2) {

  p <- plot_net(ps_list_abundant$ps[[i]], distance = "(A+B-2*J)/(A+B)", type = "taxa", 
           maxdist = 0.4, color="class", point_label="genus") +
       ggtitle(ps_list_abundant$title[[i]])

  print(p)
}

for (i in c(3)) {

  p <- plot_net(ps_list_abundant$ps[[i]], distance = "(A+B-2*J)/(A+B)", type = "taxa", 
           maxdist = 0.4, color="class", point_label="family") +
       ggtitle(ps_list_abundant$title[[i]])

  print(p)
}

```



